bootloader/lk.go

```diff
@@ -60,38 +81,189 @@ func (l *lk) Name() string {
+func (l *lk) partLabelForRoleAndTime() string {
+	// TODO: should the partition labels be fetched from gadget.yaml instead? we
+	//       have roles that we could use in the gadget.yaml structures to find
+	//       them
+	label := ""
+	switch l.role {
+	case RoleSole, RoleRunMode:
+		label = "snapbootsel"
+	case RoleRecovery:
+		label = "snaprecoverysel"
+	default:
+		panic(fmt.Sprintf("unknown bootloader role for littlekernel: %s", l.role))
+	}
+	if l.prepareImageTime {
+		// conf files at build time have .bin suffix, so technically this now
+		// becomes not a partition label but a filename, but meh names are hard
+		label += ".bin"
+	}
+	return label
+}
+
+func (l *lk) envFile() (string, error) {
 	// as for dir, we have two scenarios, image building and runtime
-	if l.inRuntimeMode {
-		// TO-DO: this should be eventually fetched from gadget.yaml
-		return filepath.Join(l.dir(), "snapbootsel")
+	partitionLabelOrConfFile := l.partLabelForRoleAndTime()
+	if l.prepareImageTime {
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
+
+	if l.role == RoleSole {
+		// see TODO: in l.dir(), this should eventually also be using
+		// envFileForPartName() too
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
+
+	// for RoleRun or RoleRecovery, we need to find the partition securely
+	envFile, _, err := l.envFileForPartName(partitionLabelOrConfFile)
+	if err != nil {
+		return "", err
 	}
-	return filepath.Join(l.dir(), "snapbootsel.bin")
+	return envFile, nil
+}
+
+// backupEnvFile returns the backup environment file, which in the case of a
+// config file is just another file, but in the runtime case is a different
+// partition.
+func (l *lk) backupEnvFile() (string, error) {
+	// for the backup file, just append "bak" to the label / file
+	partitionLabelOrConfFile := l.partLabelForRoleAndTime() + "bak"
+	if l.prepareImageTime {
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
+
+	if l.role == RoleSole {
+		// see TODO: in l.dir(), this should eventually also be using
+		// envFileForPartName() too
+		return filepath.Join(l.dir(), partitionLabelOrConfFile), nil
+	}
+
+	// for RoleRun or RoleRecovery, we need to find the partition securely
+	backupEnvFile, _, err := l.envFileForPartName(partitionLabelOrConfFile)
+	if err != nil {
+		return "", err
+	}
+	return backupEnvFile, nil
+}
+
+// envFileForPartName returns the environment file in /dev for the partition
+// name, which will always be a partition on the disk given by
+// the kernel command line parameter "snapd_lk_boot_disk" set by the bootloader.
+// It returns a boolean as the second parameter which is primarily used by
+// Present() to indicate if the searching process got "far enough" to reasonably
+// conclude that the device is using a lk bootloader, but we had errors finding
+// it. This feature is mainly for better error reporting in logs.
+func (l *lk) envFileForPartName(partName string) (string, bool, error) {
```

Comment:
again, maybe this should be envPart  and do l.partLabelForRoleAndTime() on its own and just take a flag/suffix to know if it's the backup one

Reply:
actuallly I see below that this is used for the boot partition and just the sel ones, maybe we need two level of helpers, the lowest one with a better name

Reply:
we renamed this I think it's more clear